state/src/replication/raft.rs

```diff
@@ -414,27 +434,82 @@ impl RaftClient {
         self.propose_transition(proposal).await.map(|_| id)
     }
 
-    /// Try promoting all learners if the current node is the leader
-    pub async fn try_promote_learners(&self) -> Result<(), ReplicationV2Error> {
+    /// Sync the state of the cluster as discovered through gossip with the raft
+    /// membership
+    ///
+    /// This method is called periodically by a timer and performs three sync
+    /// functions:
+    ///     - Add new learners that were missed
+    ///     - Promote any learners that are eligible to be voters
+    ///     - Expire any raft peers that are not in the list of known peers
+    pub async fn sync_membership(
+        &self,
+        known_peers: Vec<WrappedPeerId>,
+    ) -> Result<(), ReplicationV2Error> {
+        // Only the leader should update raft membership
         let leader = self.leader().await.unwrap_or(0 /* invalid id */);
         if leader != self.node_id() {
             return Ok(());
         }
 
+        // After each step we return if a change was made, only one membership change
+        // may be made at a time
+        if self.add_missed_learners(&known_peers).await? > 0 {
+            return Ok(());
+        }
+
+        if self.try_promote_learners().await? > 0 {
+            return Ok(());
+        }
+
+        self.check_expired_nodes(&known_peers).await
```

Comment:
IMO we should check for expired nodes first - in the case that there are both learners to add and peers to expire, we don't want to try replicating membership changes to nodes that are down. We also get to avoid sending a membership to change to the new learners when these peers are expired

Reply:
good point, updated